pkg/kube/krt/collection.go

```diff
@@ -0,0 +1,478 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pkg/kube/controllers"
+	istiolog "istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/maps"
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/util/sets"
+)
+
+// manyCollection builds a mapping from I->O.
+// This can be built from transformation functions of I->*O or I->[]O; both are implemented by this same struct.
+// Locking used here is somewhat complex. We use two locks, mu and recomputeMu.
+//   - mu is responsible for locking the actual data we are storing. List()/Get() calls will lock this.
+//   - recomputeMu is responsible for ensuring there is mutually exclusive access to recomputation. Typically, in a controller
+//     pattern this would be accomplished by a queue. However, these add operational and performance overhead that is not required here.
+//     Instead, we ensure at most one goroutine is recomputing things at a time.
+//     This avoids two dependency updates happening concurrently and writing events out of order.
+type manyCollection[I, O any] struct {
+	// name provides the name for this collection.
+	name string
+	// parent is the input collection we are building off of.
+	parent Collection[I]
+
+	// log is a logger for the collection, with additional labels already added to identify it.
+	log *istiolog.Scope
+
+	// recomputeMu blocks a recomputation of I->O.
+	recomputeMu sync.Mutex
+
+	// mu protects all items grouped below.
+	// This is acquired for reads and writes of data.
+	// This can be acquired with recomputeMu held, but only with strict ordering (mu inside recomputeMu)
+	mu              sync.Mutex
```

Comment:
Thoughts on an abstraction for this 2-stage locking to prevent out of order locks? I'm not convinced it would help but maybe worth a discussion.